Platforms/QemuQ35Pkg/Library/ConfigKnobShimLib/ConfigKnobShimLibCommon.c

```diff
@@ -0,0 +1,160 @@
+
+  If the config knob is not found in variable storage or the policy cache, this function returns EFI_NOT_FOUND
+  and NULL in ConfigKnobData. In this case, the autogen header will return the default value for the profile.
+
+  This function is only expected to be called from the autogen header code, all consumers of config knobs are
+  expected to use the getter functions in the autogen header.
+
+  @param[in]  ConfigKnobGuid      The GUID of the requested config knob.
+  @param[in]  ConfigKnobName      The name of the requested config knob.
+  @param[out] ConfigKnobData      The retrieved data of the requested config knob. The caller will allocate memory
+                                  for this buffer and the caller is responsible for freeing it.
+  @param[in] ConfigKnobDataSize   The allocated size of ConfigKnobData. This will be set to the correct value for the
+                                  size of ConfigKnobData in the success case. This should equal ProfileDefaultSize, if
+                                  not, EFI_BAD_BUFFER_SIZE will be returned.
+  @param[in] ProfileDefaultValue  The profile defined default value of this knob, to be filled in if knob is not
+                                  overridden in a cached policy or variable storage.
+
+  @retval EFI_INVALID_PARAMETER   Input argument is null.
+  @retval EFI_BAD_BUFFER_SIZE     Caller passed in a value of ConfigKnobDataSize that does not equal
+                                  ProfileDefaultSize.
+  @retval EFI_SUCCESS             The operation succeeds.
+
+**/
+STATIC
+EFI_STATUS
+GetConfigKnob (
+  IN EFI_GUID  *ConfigKnobGuid,
+  IN CHAR16    *ConfigKnobName,
+  OUT VOID     *ConfigKnobData,
+  IN UINTN     ConfigKnobDataSize,
+  IN VOID      *ProfileDefaultValue
+  )
+{
+  EFI_STATUS  Status;
+  UINTN       VariableSize = ConfigKnobDataSize;
+
+  if ((ConfigKnobGuid == NULL) || (ConfigKnobName == NULL) || (ConfigKnobData == NULL) ||
+      (ConfigKnobDataSize == 0) || (ProfileDefaultValue == NULL))
+  {
+    DEBUG ((DEBUG_ERROR, "%a: Invalid parameter!\n", __FUNCTION__));
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  // Check if it is in variable storage
+  Status = GetConfigKnobFromVariable (ConfigKnobGuid, ConfigKnobName, ConfigKnobData, &VariableSize);
+
+  if (ConfigKnobDataSize != VariableSize) {
+    // we will only accept this variable if it is the correct size
+    Status = EFI_NOT_FOUND;
+  }
+
+  if (EFI_ERROR (Status)) {
+    // we didn't find the value in variable storage, which is expected if the knob has not been overridden, or
+    // the size mismatched. In either case, we default to the profile value. This is a decision the platform may
+    // choose not to take if it does not want to risk having different components having different values for config
+    // knobs. Instead, this function could fail if the knob is not found (but still return the profile default if
+    // the size mismatched)
+    DEBUG ((
+      DEBUG_VERBOSE,

```

Comment:
Probably worth at a warning level?

Reply:
I left this as verbose, because it will print for every non-overriden config knob (it didn't find it in var storage), which may be useful for debugging, but could spam the log otherwise. I added a warning for the case where a variable was found but the size mismatched.